sdk/nodejs/provisioning/docker-provision/image.ts

```diff
@@ -0,0 +1,237 @@
+
+	/**
+	 * trim image digests to 16 characters to make output more readable.
+	 */
+	private readonly DIGEST_LEN = 16;
+
+	constructor(ref: string) {
+		// Throw error if ref is not correctly formatted.
+		ImageRef.validate(ref);
+
+		this.ref = ref;
+
+		const id = ref.split('@sha256:', 2)[1];
+		this.id = id.slice(0, this.DIGEST_LEN);
+	}
+
+	get Ref(): string {
+		return this.ref;
+	}
+
+	get ID(): string {
+		return this.id;
+	}
+
+	/**
+	 * validateImage verify that the passed ref
+	 * is compliant with DockerImage constructor.
+	 *
+	 * This function does not return anything but
+	 * only throw on error.
+	 *
+	 * @throws no digest found in ref.
+	 */
+	static validate(ref: string): void {
+		if (!ref.includes('@sha256')) {
+			throw new Error(`no digest found in ref ${ ref }`);
+		}
+	}
+}
+
+/**
+ * DockerImage is an implementation of EngineConn to set up a Dagger
+ * Engine session from a pulled docker image.
+ */
+export class DockerImage implements EngineConn {
+	private imageRef: ImageRef;
+
+	private readonly cacheDir = path.join(
+		process.env.XDG_CACHE_HOME || path.join(os.homedir(), '.cache'),
+		'dagger'
+	);
+
+	private readonly ENGINE_SESSION_BINARY_PREFIX = 'dagger-engine-session';
+
+	private enginePid?: number;
+
+	constructor(u: URL) {
+		this.imageRef = new ImageRef(u.host + u.pathname);
+	}
+
+	Addr(): string {
+		return 'http://dagger';
+	}
+
+	async Connect(opts: ConnectOpts): Promise<Client> {
+		this.createCacheDir();
+
+		const engineSessionBinPath = this.buildBinPath();
+		if (!fs.existsSync(engineSessionBinPath)) {
+			this.pullEngineSessionBin(engineSessionBinPath);
+		}
+
+		return this.runEngineSession(engineSessionBinPath, opts);
+	}
+
+	/**
+	 * createCacheDir will create a cache directory on user
+	 * host to store dagger binary.
+	 *
+	 * If set, it will use XDG directory, if not, it will use `$HOME/.cache`
+	 * as base path.
+	 * Nothing happens if the directory already exists.
+	 */
+	private createCacheDir(): void {
+		fs.mkdirSync(this.cacheDir, { mode: 0o700, recursive: true });
+	}
+
+	/**
+	 * buildBinPath create a path to output engine session binary.
+	 *
+	 * It will store it in the cache directory with a name composed
+	 * of the base engine session as constant and the engine identifier.
+	 */
+	private buildBinPath(): string {
+		const binPath = `${ this.cacheDir }/${ this.ENGINE_SESSION_BINARY_PREFIX }-${ this.imageRef.ID }`;
+
+		switch (os.platform()) {
+			case 'win32':
+				return `${ binPath }.exe`;
+			default:
+				return binPath;
+		}
+	}
+
+	/**
+	 * pullEngineSessionBin will retrieve Dagger binary from its docker image
+	 * and copy it to the local host.
+	 * This function automatically resolves host's platform to copy the correct
+	 * binary.
+	 */
+	private pullEngineSessionBin(engineSessionBinPath: string): void {
+		// Create a temporary bin file path
+		const tmpBinPath = path.join(
+			this.cacheDir,
+			`temp-${ this.ENGINE_SESSION_BINARY_PREFIX }`
```

Comment:
Just noticed we need to make this an actual random file (e.g. `temp-dagger-engine-session-<randomid>` is what we do in other sdks), otherwise if any other provisioning code is running in parallel with this one then they will interfere with one another.

It's generally best to use the stdlib for this (what we do in go+python) but I guess nodejs only gives you the ability to make a temporary directory: https://nodejs.org/api/fs.html#fsmkdtempprefix-options-callback

We can use that though, just need to make that temp directory inside the cache dir here, then can create that tempdir, put `temp-dagger-engine-session` in there and then rename it to the final binary. Just be sure to remove the tempdir when we're done.

Reply:
An even simpler possibility would be to use unix nanosecond timestamp as the <randomid>. That reduces the chances of a collision to be so small that it's probably negligible. Just in case that makes this easier to update.

Reply:
Done by using a random unix timestamp in nanosecond.